Userland/Libraries/LibAudio/PlaybackStreamSerenity.cpp

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2023, Gregory Bertilson <zaggy1024@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "PlaybackStreamSerenity.h"
+
+#include <LibCore/ThreadedPromise.h>
+
+namespace Audio {
+
+ErrorOr<NonnullRefPtr<PlaybackStream>> PlaybackStreamSerenity::create(OutputState initial_state, u32 sample_rate, u8 channels, u32 target_latency_ms, AudioDataRequestCallback&& data_request_callback)
+{
+    VERIFY(data_request_callback);
+    auto connection = TRY(ConnectionToServer::try_create());
+    if (auto result = connection->try_set_self_sample_rate(sample_rate); result.is_error())
+        return Error::from_string_literal("Failed to set sample rate");
+    (void)channels, (void)target_latency_ms;
+
+    auto polling_timer = TRY(Core::Timer::try_create());
+    auto implementation = TRY(adopt_nonnull_ref_or_enomem(new (nothrow) PlaybackStreamSerenity(connection, move(polling_timer), move(data_request_callback))));
+    if (initial_state == OutputState::Playing)
+        connection->async_start_playback();
+    return implementation;
+}
+
+PlaybackStreamSerenity::PlaybackStreamSerenity(NonnullRefPtr<ConnectionToServer> stream, NonnullRefPtr<Core::Timer> polling_timer, AudioDataRequestCallback&& data_request_callback)
+    : m_connection(move(stream))
+    , m_polling_timer(move(polling_timer))
+    , m_data_request_callback(move(data_request_callback))
+{
+    // Ensure that our audio buffers are filled when they are more than 3/4 empty.
+    // FIXME: Add an event to ConnectionToServer track the sample rate and update this interval.
+    //        Or use a shared condition variable to wake an enqueuing thread to provide data to the audio server.
+    m_polling_timer->set_interval(static_cast<int>((AUDIO_BUFFERS_COUNT * 3 / 4) * AUDIO_BUFFER_SIZE * 1000 / m_connection->get_self_sample_rate()));
+    m_polling_timer->on_timeout = [this]() {
+        fill_buffers();
+    };
+    m_polling_timer->start();
+}
+
+void PlaybackStreamSerenity::fill_buffers()
+{
+    // The number below is picked arbitrarily.
+    // FIXME: This shouldn't need to poll. Ideally, we should be able to recieve a signal when the server
+    //        consumes a buffer and enqueue some data automatically.
+    while (m_connection->can_enqueue()) {
+        Array<Sample, AUDIO_BUFFER_SIZE> buffer;
+        buffer.fill({ 0.0f, 0.0f });
+        auto written_data = m_data_request_callback(Bytes { reinterpret_cast<u8*>(buffer.data()), sizeof(buffer) }, PcmSampleFormat::Float32, AUDIO_BUFFER_SIZE);
+        // FIXME: We should make it so that we can enqueue the exact number of samples we receive from the callback.
+        //        Otherwise, the number of samples that are reported by the server will not be correct for the
+        //        client of this implementation. Eventually, we will want to use server-side timing information, which
+        //        should be accurate down to single samples.
```

Comment:
That means splitting out the UserSampleQueue into a reusable non-realtime frontend for multiple clients, can you make this comment more specific? (Server-side timing as well, yes)

Reply:
This isn't so much related to `UserSampleQueue` as it is to the `AudioServer::ClientAudioStream`. The idea is that the total number of samples the server sees will be a multiple of `AUDIO_BUFFER_SIZE` rather than being the exact number of samples the client has written into those buffers. In the process of implementing timing information, I forsee having to change that so that we can track the maximum number of samples that we can count up to when interpolating the current playback time. Unsure whether that _has_ to be implemented on the server side, but I'm pretty sure that would be the simplest solution.

I've updated this comment as well.